catroid/src/main/java/org/catrobat/catroid/devices/arduino/ArduinoImpl.java

```diff
@@ -225,27 +229,27 @@ public void setAnalogArduinoPin(int pin, int value) {
 	@Override
 	public void setDigitalArduinoPin(int digitalPinNumber, int pinValue) {
 		int digitalPort = 0;
-		double value;
+		int PinNumberOfPort;
+		int PortValue;
+
 		if (digitalPinNumber < 8) {
-			if (pinValue > 0) {
-				value = Math.pow(2, (double) digitalPinNumber);
-				sendDigitalFirmataMessage(digitalPort, digitalPinNumber, (int) value);
-				arduinoListener.setPortValue(digitalPinNumber, 1);
-			} else {
-				sendDigitalFirmataMessage(digitalPort, digitalPinNumber, 0);
-				arduinoListener.setPortValue(digitalPinNumber, 0);
-			}
+			digitalPort = 0;
+			PinNumberOfPort = digitalPinNumber;
 		} else {
 			digitalPort = 1;
-			if (pinValue > 0) {
-				value = Math.pow(2, (double) digitalPinNumber - 8);
-				sendDigitalFirmataMessage(digitalPort, digitalPinNumber, (int) value);
-				arduinoListener.setPortValue(digitalPinNumber, 1);
-			} else {
-				sendDigitalFirmataMessage(digitalPort, digitalPinNumber, 0);
-				arduinoListener.setPortValue(digitalPinNumber, 0);
-			}
+			PinNumberOfPort = digitalPinNumber - 8;
+		}
+
+		PortValue = arduinoListener.getuCPortValue(digitalPort);
+		if (pinValue > 0) { // set pin
+			PortValue = PortValue | (1 << PinNumberOfPort);
+			arduinoListener.setPortValue(digitalPinNumber, 1);
+		} else { // clear pin
+			PortValue = PortValue & ~(1 << PinNumberOfPort);
```

Comment:
If you want, you could create a util method to set a bit. Like
```java
public static int setBit(int number, int index, int value) {
  if (value == 0) {
    return number & ~(1 << index);
  } else {
    return number | (1 << index);
  }
}
```
When using the method somewhere in the code, the reader immediately knows what it does. Then you can also remove the comments here.